File: engine/runtime/src/main/scala/org/enso/compiler/pass/optimise/UnreachableMatchBranches.scala
package org.enso.compiler.pass.optimise

import org.enso.compiler.context.{InlineContext, ModuleContext}
import org.enso.compiler.core.IR
import org.enso.compiler.core.IR.{IdentifiedLocation, Pattern}
import org.enso.compiler.exception.CompilerError
import org.enso.compiler.pass.IRPass
import org.enso.compiler.pass.analyse.{
  AliasAnalysis,
  DataflowAnalysis,
  DemandAnalysis,
  TailCall
}
import org.enso.compiler.pass.desugar._
import org.enso.compiler.pass.resolve.{DocumentationComments, IgnoredBindings}
import org.enso.syntax.text.Location

import scala.annotation.unused

/** This pass discovers and optimised away unreachable case branches.
  *
  * It removes these unreachable expressions from the IR, and attaches a
  * [[IR.Warning]] diagnostic to the case expression itself.
  *
  * Currently, a branch is considered 'unreachable' by this pass if:
  *
  * - It occurs after a catch-all branch.
  *
  * In the future, this pass should be expanded to consider patterns that are
  * entirely subsumed by previous patterns in its definition of uncreachable,
  * but this requires doing sophisticated coverage analysis, and hence should
  * happen as part of the broader refactor of nested patterns desugaring.
  *
  * This pass requires no configuration.
  *
  * This pass requires the context to provide:
  *
  * - Nothing
  */
case object UnreachableMatchBranches extends IRPass {
  override type Metadata = IRPass.Metadata.Empty
  override type Config   = IRPass.Configuration.Default

  override val precursorPasses: Seq[IRPass] = List(
    ComplexType,
    DocumentationComments,
    FunctionBinding,
    GenerateMethodBodies,
    LambdaShorthandToLambda
  )
  override val invalidatedPasses: Seq[IRPass] = List(
    AliasAnalysis,
    DataflowAnalysis,
    DemandAnalysis,
    IgnoredBindings,
    NestedPatternMatch,
    TailCall
  )

  /** Runs unreachable branch optimisation on a module.
    *
    * @param ir the Enso IR to process
    * @param moduleContext a context object that contains the information needed
    *                      to process a module
    * @return `ir`, possibly having made transformations or annotations to that
    *         IR.
    */
  override def runModule(
    ir: IR.Module,
    @unused moduleContext: ModuleContext
  ): IR.Module = {
    ir.mapExpressions(optimizeExpression)
  }

  /** Runs unreachable branch optimisation on an expression.
    *
    * @param ir the Enso IR to process
    * @param inlineContext a context object that contains the information needed
    *                      for inline evaluation
    * @return `ir`, possibly having made transformations or annotations to that
    *         IR.
    */
  override def runExpression(
    ir: IR.Expression,
    @unused inlineContext: InlineContext
  ): IR.Expression = {
    ir.transformExpressions { case x =>
      optimizeExpression(x)
    }
  }

  /** @inheritdoc */
  override def updateMetadataInDuplicate[T <: IR](
    @unused sourceIr: T,
    copyOfIr: T
  ): T = copyOfIr

  // === Pass Internals =======================================================

  /** Optimizes an expression by removing unreachable branches in case
    * expressions.
    *
    * @param expression the expression to optimize
    * @return `expression` with unreachable case branches removed
    */
  def optimizeExpression(expression: IR.Expression): IR.Expression = {
    expression.transformExpressions { case cse: IR.Case =>
      optimizeCase(cse)
    }
  }

  /** Optimizes a case expression by removing unreachable branches.
    *
    * Additionally, it will attach a warning about unreachable branches to the
    * case expression.
    *
    * @param cse the case expression to optimize
    * @return `cse` with unreachable branches removed
    */
  //noinspection DuplicatedCode
  def optimizeCase(cse: IR.Case): IR.Case = {
    cse match {
      case expr @ IR.Case.Expr(scrutinee, branches, _, _, _) =>
        val reachableNonCatchAllBranches = branches.takeWhile(!isCatchAll(_))
        val firstCatchAll                = branches.find(isCatchAll)
        val unreachableBranches =
          branches.dropWhile(!isCatchAll(_)).drop(1)
        val reachableBranches = firstCatchAll
          .flatMap(b => Some(reachableNonCatchAllBranches :+ b))
          .getOrElse(List())
          .toList

        if (unreachableBranches.isEmpty) {
          expr.copy(
            scrutinee = optimizeExpression(scrutinee),
            branches = branches.map(b =>
              b.copy(expression = optimizeExpression(b.expression))
            )
          )
        } else {
          val unreachableLocation =
            unreachableBranches.foldLeft(None: Option[IdentifiedLocation])(
              (loc, branch) => {
                loc match {
                  case Some(loc) =>
                    branch.location match {
                      case Some(branchLoc) =>
                        Some(
                          IdentifiedLocation(
                            Location(loc.start, branchLoc.end),
                            loc.id
                          )
                        )
                      case None => Some(loc)
                    }
                  case None => branch.location
                }
              }
            )

          val diagnostic = IR.Warning.Unreachable.Branches(unreachableLocation)

          expr
            .copy(
              scrutinee = optimizeExpression(scrutinee),
              branches = reachableBranches
                .map(b => b.copy(expression = optimizeExpression(b.expression)))
            )
            .addDiagnostic(diagnostic)
        }
      case _: IR.Case.Branch =>
        throw new CompilerError("Unexpected case branch.")
    }
  }

  /** Determines if a branch is a catch all branch.
    *
    * @param branch the branch to check
    * @return `true` if `branch` is catch-all, otherwise `false`
    */
  def isCatchAll(branch: IR.Case.Branch): Boolean = {
    branch.pattern match {
      case _: Pattern.Name          => true
      case _: Pattern.Constructor   => false
      case _: Pattern.Documentation => false
      case _: IR.Error.Pattern      => true
    }
  }
}

File: engine/runtime-compiler/src/main/java/org/enso/compiler/pass/optimise/UnreachableMatchBranches.java
package org.enso.compiler.pass.optimise;

import java.util.ArrayList;
import java.util.stream.Stream;
import org.enso.compiler.context.InlineContext;
import org.enso.compiler.context.ModuleContext;
import org.enso.compiler.core.CompilerError;
import org.enso.compiler.core.ir.Expression;
import org.enso.compiler.core.ir.IdentifiedLocation;
import org.enso.compiler.core.ir.Pattern;
import org.enso.compiler.core.ir.expression.Case;
import org.enso.compiler.core.ir.expression.warnings.Unreachable;
import org.enso.compiler.pass.IRProcessingPass;
import org.enso.compiler.pass.MiniIRPass;
import org.enso.compiler.pass.MiniPassFactory;
import org.enso.compiler.pass.analyse.AliasAnalysis$;
import org.enso.compiler.pass.analyse.DataflowAnalysis$;
import org.enso.compiler.pass.analyse.DemandAnalysis$;
import org.enso.compiler.pass.analyse.TailCall;
import org.enso.compiler.pass.desugar.ComplexType$;
import org.enso.compiler.pass.desugar.FunctionBinding$;
import org.enso.compiler.pass.desugar.GenerateMethodBodies$;
import org.enso.compiler.pass.desugar.LambdaShorthandToLambda$;
import org.enso.compiler.pass.desugar.NestedPatternMatch$;
import org.enso.compiler.pass.resolve.DocumentationComments$;
import org.enso.compiler.pass.resolve.IgnoredBindings$;
import org.enso.scala.wrapper.ScalaConversions;
import scala.collection.immutable.List;
import scala.jdk.javaapi.CollectionConverters;

/**
 * This pass discovers and optimizes away unreachable case branches.
 *
 * <p>It removes these unreachable expressions from the IR, and attaches a {@link
 * org.enso.compiler.core.ir.Warning} diagnostic to the case expression itself.
 *
 * <p>Currently, a branch is considered 'unreachable' by this pass if:
 *
 * <ul>
 *   <li>It occurs after a catch-all branch.
 * </ul>
 *
 * <p>In the future, this pass should be expanded to consider patterns that are entirely subsumed by
 * previous patterns in its definition of unreachable, but this requires doing sophisticated
 * coverage analysis, and hence should happen as part of the broader refactor of nested patterns
 * desugaring.
 *
 * <p>This pass requires no configuration.
 *
 * <p>This pass requires the context to provide:
 *
 * <ul>
 *   <li>Nothing
 * </ul>
 */
public final class UnreachableMatchBranches implements MiniPassFactory {
  private UnreachableMatchBranches() {}

  public static final UnreachableMatchBranches INSTANCE = new UnreachableMatchBranches();

  @Override
  public List<IRProcessingPass> precursorPasses() {
    java.util.List<IRProcessingPass> passes = new ArrayList<>();
    passes.add(ComplexType$.MODULE$);
    passes.add(DocumentationComments$.MODULE$);
    passes.add(FunctionBinding$.MODULE$);
    passes.add(GenerateMethodBodies$.MODULE$);
    passes.add(LambdaShorthandToLambda$.MODULE$);
    return CollectionConverters.asScala(passes).toList();
  }

  @Override
  public List<IRProcessingPass> invalidatedPasses() {
    java.util.List<IRProcessingPass> passes = new ArrayList<>();
    passes.add(AliasAnalysis$.MODULE$);
    passes.add(DataflowAnalysis$.MODULE$);
    passes.add(DemandAnalysis$.MODULE$);
    passes.add(IgnoredBindings$.MODULE$);
    passes.add(NestedPatternMatch$.MODULE$);
    passes.add(TailCall.INSTANCE);
    return CollectionConverters.asScala(passes).toList();
  }

  @Override
  public MiniIRPass createForInlineCompilation(InlineContext inlineContext) {
    return new Mini();
  }

  @Override
  public MiniIRPass createForModuleCompilation(ModuleContext moduleContext) {
    return new Mini();
  }

  private static class Mini extends MiniIRPass {
    @Override
    public Expression transformExpression(Expression expr) {
      return switch (expr) {
        case Case cse -> optimizeCase(cse);
        default -> expr;
      };
    }

    /**
     * Optimizes a case expression by removing unreachable branches.
     *
     * <p>Additionally, it will attach a warning about unreachable branches to the case expression.
     *
     * @param cse the case expression to optimize
     * @return `cse` with unreachable branches removed
     */
    private Case optimizeCase(Case cse) {
      if (cse instanceof Case.Expr expr) {
        var branches = CollectionConverters.asJava(expr.branches());
        var reachableNonCatchAllBranches =
            branches.stream().takeWhile(branch -> !isCatchAll(branch));
        var firstCatchAll = branches.stream().filter(this::isCatchAll).findFirst();
        var unreachableBranches =
            branches.stream().dropWhile(branch -> !isCatchAll(branch)).skip(1).toList();
        List<Case.Branch> reachableBranches;
        if (firstCatchAll.isPresent()) {
          reachableBranches = appended(reachableNonCatchAllBranches, firstCatchAll.get());
        } else {
          reachableBranches = ScalaConversions.nil();
        }

        if (unreachableBranches.isEmpty()) {
          return expr;
        } else {
          var firstUnreachableWithLoc =
              unreachableBranches.stream()
                  .filter(branch -> branch.identifiedLocation() != null)
                  .findFirst();
          var lastUnreachableWithLoc =
              unreachableBranches.stream()
                  .filter(branch -> branch.identifiedLocation() != null)
                  .reduce((first, second) -> second);
          IdentifiedLocation unreachableLocation = null;
          if (firstUnreachableWithLoc.isPresent() && lastUnreachableWithLoc.isPresent()) {
            unreachableLocation =
                new IdentifiedLocation(
                    firstUnreachableWithLoc.get().location().get().start(),
                    lastUnreachableWithLoc.get().location().get().end(),
                    firstUnreachableWithLoc.get().id());
          }

          var diagnostic = new Unreachable.Branches(unreachableLocation);
          var copiedExpr = expr.copyWithBranches(reachableBranches);
          copiedExpr.getDiagnostics().add(diagnostic);
          return copiedExpr;
        }
      } else {
        throw new CompilerError("Unexpected case branch.");
      }
    }

    /**
     * Determines if a branch is a catch all branch.
     *
     * @param branch the branch to check
     * @return `true` if `branch` is catch-all, otherwise `false`
     */
    private boolean isCatchAll(Case.Branch branch) {
      return switch (branch.pattern()) {
        case Pattern.Name ignored -> true;
        default -> false;
      };
    }

    private static List<Case.Branch> appended(Stream<Case.Branch> branches, Case.Branch branch) {
      var ret = new ArrayList<>(branches.toList());
      ret.add(branch);
      return CollectionConverters.asScala(ret).toList();
    }
  }
}
